### PR TITLE
Make Ray releases tests use `pydantic>=2.5.0`

### PR DESCRIPTION
**Pull request by can-anyscale**
## Why are these changes needed?

As a follow-up to #40451, this change makes all the Ray release tests use `pydantic>=2.5.0`.

**Additional changes:**
* The linter now uses `mypy==1.7.0`.
* All Ray release tests, including workspace templates, now use `deepspeed>=0.12.3`.
* This change fix the `serve_resnet_benchmark.py` release test, which is broken on `master`.

## Checks

- [X] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [X] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [X] Release tests
     - This change updates Ray release tests.

### CODE REVIEW

**Review by aslonnie — COMMENTED** [review]
is this ready for review? could you say more in the PR title and description for the context?

**Comment by can-anyscale** [comment]
w00t did I add you as a reviewer; this is for @shrekris-anyscale , he will pick this up

File: release/requirements_buildkite.txt [hunk 1]
```diff
@@ -1,5 +1,5 @@
 #
-# This file is autogenerated by pip-compile with python 3.7
+# This file is autogenerated by pip-compile with python 3.9
```
Comment (can-anyscale) [inline release/requirements_buildkite.txt:2]: should be generated with python 3.7

File: release/requirements_buildkite.in [hunk 2]
```diff
@@ -8,7 +9,7 @@ freezegun
 google-cloud-storage
 jinja2
 protobuf >= 3.15.3, != 3.19.5
-pydantic < 1.10.0
+pydantic >= 2.5.0
```
Comment (can-anyscale) [inline release/requirements_buildkite.in:12]: or just remove pydantic i think it's no longer a requirement

File: release/requirements_buildkite.txt (hunk 1 above)
Comment (shrekris-anyscale) [reply release/requirements_buildkite.txt]: Oh that makes sense. I'll update it.

File: release/requirements_buildkite.in (hunk 2 above)
Comment (shrekris-anyscale) [reply release/requirements_buildkite.in:11]: Should I also remove `anyscale`?

**Comment by shrekris-anyscale** [comment]
`serve_resnet_benchmark` passed with the change. [Buildkite link](https://buildkite.com/ray-project/release/builds/1695#018be95c-82be-410f-9e38-5589409eb77d)

**Comment by shrekris-anyscale** [comment]
The only remaining stable/non-flaky/non-jailed test that failed is `long_running_serve.aws` ([Buildkite link](https://buildkite.com/ray-project/release/builds/1672#018be3bd-f4b1-4fe3-8da9-6e96e8b20f8c)). This also failed on `master` last night due to an infra error ([link](https://buildkite.com/ray-project/release/builds/1661#018be237-2ee5-4d27-b1e0-0faa47925b65)). I don’t think this test uses pydantic in any unique way, and the error traceback was the same on master and on the PR, so I think the PR run also failed due to an infra error.

**Comment by shrekris-anyscale** [comment]
Here's [a link to the Buildkite run](https://buildkite.com/ray-project/release/builds/1672#_) where the rest of the stable/non-flaky/non-jailed tests pass.

**Review by aslonnie — APPROVED** [review]

**Review by ericl — APPROVED** [review]
Stamping for codeowners

**Comment by shrekris-anyscale** [comment]
I kicked off a `long_running_serve.aws` run on [the commit](https://github.com/ray-project/ray/pull/41237) just before the pydantic version was updated: [Buildkite link](https://buildkite.com/ray-project/release/builds/1578).

**Review by matthewdeng — APPROVED** [review]

**Review by kouroshHakha — APPROVED** [review]
I checked the failures on RLlib. They are not related. Separately pinged @sven1977 on what needs to be done there.

**Review by c21 — APPROVED** [review]
The jailed test failure on Data side should be irrelevant (out of disk error). Okay to merge.

**Comment by shrekris-anyscale** [comment]
The `long_running_serve.aws` smoke test ran successfully: [Buildkite link](https://buildkite.com/ray-project/release/builds/1775#018bef0f-30bc-4080-ae6d-a7ebeae2a235)

**Comment by shrekris-anyscale** [comment]
> I kicked off a long_running_serve.aws run on https://github.com/ray-project/ray/pull/41237 just before the pydantic version was updated: [Buildkite link](https://buildkite.com/ray-project/release/builds/1578).

The run failed with the same error as this PR. It's unlikely that this PR is the root cause. There are no remaining test failures connected to this PR.

@architkulkarni This PR is ready to merge.

**Comment by edoakes** [comment]
Thanks for helping out with this @can-anyscale

**Pull request merged**